### ISSUE TITLE
JS files chache

### ISSUE THREAD

**Issue by mariusrak** [arangodb/arangodb#1723]
Hello, i have this entry in my _system.routing database:
`{"url":"/db","action":{"controller":"D:\\Node\\...\\db_server\\server.js"}}`
Which means i can write JS code into the server.js file and it's run when i hit `http://localhost:8529/db`
However i encounter big problems with cache. When I update the file, it is not reloaded. Running `require('internal').reloadRouting();` doesn't have any effect.

I edited source code at C:\Program Files\ArangoDB 2.8\share\arangodb\js\server\modules\org\arangodb\actions.js:411 and added

    if(action.hasOwnProperty('no_cache')){
      if(context.module._cache){
        context.module._cache = {};
      }
      if(context.module.cache){
        context.module.cache = {};
      }
    }

And updated the entry in database to 
`{"url":"/db","action":{"no_cache":true,"controller":"D:\\Node\\...\\db_server\\server.js"}}`
Now everything works as expected. Though I still must reloadRouting, but now it works.

Is there any way to do this some legitimate way? Or is this my 'no_cache' hack some missing future? Would it be possible to add some mechanism to disable cache on demand to future versions?
Thanks

**Comment by pluma** [comment]
Hi @mariusrak,

you are correct that modules are cached independently of the routing cache. Clearing the module cache (or preventing a module from being cached) is currently not supported.

The actions mechanism is really only intended as an internal API and only supported for backwards compatibility with early ArangoDB versions and some edge cases.

As you may have noticed while digging through the ArangoDB source code, Foxx provides a per-service module cache which is cleared whenever a Foxx service is reloaded. I would strongly encourage you to see whether Foxx fits your use case before continuing to dig into the actions mechanism.

It's actually possible to create a Foxx service with just two files (a manifest and a controller file) and without using repositories or models (you can just use the same APIs available in actions).

You just need a controller file like this (e.g. `ctrl.js`):

```js
const Foxx = require('org/arangodb/foxx');
const ctrl = new Foxx.Controller(applicationContext);
ctrl.get('/', function (req, res) {
  res.send('Hello World');
});
```

with a `manifest.json` like this:

```json
{
  "name": "my-foxx",
  "version": "0.0.0",
  "controllers": "ctrl.js",
  "defaultDocument": "",
  "engines": {"arangodb": "^2.8.0"}
}
```

You can then mount the service (upload a zip bundle) at a path like `/db` and access it:

`curl http://localhost:8529/_db/_system/db`

The upcoming 3.0 release will remove a lot of the existing conceptual overhead of Foxx which will hopefully make it even easier to get started with it.

**Comment by mariusrak** [comment]
I did as You wrote and the url redirects brower to /index.html which returns

    {"error":true,"code":404,"errorNum":404,"errorMessage":"unknown path '/db/index.html'"}

I also used "Try it out!" button in web admin.

**Comment by pluma** [comment]
@mariusrak that would indicate the "defaultDocument" setting in the manifest is missing or not set to an empty string. In ArangoDB 2.x a redirect from the index route (`/`) to `index.html` is automatically created unless that setting is provided and set to an empty string. This behaviour is confusing and will be removed in 3.0 for exactly this reason.

**Comment by mariusrak** [comment]
I just copied code you provided in your response above and it behaves the described way.

**Comment by pluma** [comment]
Funky. I'll check what's wrong.

**Comment by pluma** [comment]
Okay, it works fine for me on ArangoDB 2.8.

Here's the bundle I tried it with: [ze-app.zip](https://github.com/arangodb/arangodb/files/128006/ze-app.zip)

**Comment by mariusrak** [comment]
Your zip works fine.

Is there any way to set custom absolute path other than at arango installation directory?

**Comment by pluma** [comment]
@mariusrak try diffing the manifest bundled in the zip file against the one you created, or just work from the one in the bundle if you can't figure out what's wrong with yours.

You can only specify the mount path at "install time". It is used for storing the service on the file system and to generate the prefix for prefixed collections.

**Comment by mariusrak** [comment]
If I understand it right, mount path is part of url address to run application. Is it so?
But I want to change path where i store scripts, so I can have them in one folder with other - front end files of project. Is it possible?

**Comment by pluma** [comment]
@mariusrak you mean the folder containing the Foxx service's files (i.e. JS and manifest)?

You can keep the source files wherever you want. However you should think of mounted/installed services in ArangoDB as "deployed" and shouldn't manipulate them directly except using the development mode.

There is no way to store individual services in arbitrary locations but you can change the common root folder of all Foxx services using the `javascript.app-path` configuration option of ArangoDB. ArangoDB creates a folder structure based on the mount points of the services within that folder.

**Comment by mariusrak** [comment]
Since I'm currently using ArangoDB just for one project it isn't problem to set app-path to be located in directory of the project.

However it'd be nice in future to have possibility to have source js files stored outside of database directory so you can edit them "live"without need to upload them to see changes.

For now thanks for your advise.

**Comment by pluma** [comment]
@mariusrak if you enable development mode (click the big button in the service detail view of the web interface) you'll be shown the absolute file system location of the service and the service's modules will not be cached (or rather: the cache will be reset for every request) until the development is disabled again. This lets you edit the files on disk without needing to re-upload them.

Moving the install paths out of the app-path will not be supported within the foreseeable future due to security and consistency concerns.

**Comment by mariusrak** [comment]
Okay and is it possible to set url in production mode for a foxx app to something else as the one predefined by system. Now I have http://localhost:8529/_db/my_db_name/foxx-app/ which unveils name of database and also doesn't look nice. Is it posibble to make e.g all request received at http://localhost:8529/ go directly to http://localhost:8529/_db/organize_me/foxx-app/ ?
Thanks in advance.

**Comment by pluma** [comment]
You can use a reverse proxy like nginx or Apache httpd to restrict access to a subsection of the URL as you would with any other web application. That's outside of the scope of ArangoDB at this moment.

**Comment by mariusrak** [comment]
But it would mean that I loose the speed benefit of foxx.

**Comment by pluma** [comment]
@mariusrak no, it would add a minimal overhead to requests as the forward web server would pass the incoming requests to ArangoDB and pass on the outgoing response.
![eli5](https://cloud.githubusercontent.com/assets/1280922/13012838/3f214f76-d1ad-11e5-9d8d-5ef2ddf75dc8.png)

**Comment by pekeler** [comment]
@mariusrak the speed benefit comes from your code running in the same process as your data. When fronting ArangoDB with nginx, you can think of nginx as part of the network.

**Comment by pluma** [comment]
FWIW, it's probably a good idea to put Foxx services behind a reverse proxy in production anyway, if only to reduce the attack surface of your application. Also in some cases you may want to put additional applications like load balancers or caches in between.

**Comment by pekeler** [comment]
I'm using nginx in development because I want it to be similar to production. The important part of my nginx.conf is

```
    location / {
      root      ./foxx/_db/mydb/myapp/APP/public;
      try_files /$uri /index.html;
    }
```

At the top level of my project I have a package.json with a script-entry that starts both arangod and nginx, so it's not much of a hassle.

**Comment by mariusrak** [comment]
Okay, thanks very much guys.

**Comment by pluma** [comment]
For the record, this is what the equivalent would look like without Foxx (or just with an external application server really):

![lolstack](https://cloud.githubusercontent.com/assets/1280922/13013619/7980725c-d1b0-11e5-9672-f393eb0382f0.png)
